### PR TITLE
HDDS 5781. Enable ACLs and support for all s3 file operations.

### PR DESCRIPTION
**Pull request by neils-dev**
… error responses to client on error (fixing general 500 response in current gateway).

## What changes were proposed in this pull request?

S3 gateway Gprc support for all ozone s3 supported client requests with security and ACLs enabled.  This PR adds s3g Grpc acl support with patches to the `OMClientRequest` user thread context and to the `ozone manager`.  

Included in this PR are enhancements to the s3 gateway client to return meaningful error responses, http return codes and descriptions, to the caller on error.  For example a user accessing a resource **_without_** permission is returned `HTTP 403` `AccessDenied`.  Previously all errors on client creation return `HTTP 500`.

All secure cluster smoke tests are re-enabled in this PR.  Secure cluster is updated to enable s3g Grpc and to enable all s3 gateway acceptance tests.
 
## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5781

## How was this patch tested?
Manual Testing with secure ozone cluster:
s3 request - ozone cluster processing s3 create bucket request with permission _and without_ permission.

With secret key:
`$ cd hadoop-ozone/dist/target/ozone-1.2.0-SNAPSHOT/compose/ozonesecure`
`$ docker-compose up -d --scale datanode=3`
`$ docker-compose run scm bash`
`bash-4.2$ kinit -kt /etc/security/keytabs/testuser.keytab testuser/scm`
`bash-4.2$ ozone s3 getsecret`
`$ export AWS_ACCESS_KEY=testuser/scm@EXAMPLE.COM` `AWS_SECRET_KEY=<key>`
`$ aws s3api --endpoint http://localhost:9878 create-bucket --bucket=bucket1`
_{
"Location": "http://localhost:9878/bucket1"
}_

Without secret key:
`$ export AWS_ACCESS_KEY=`
`$ export AWS_SECRET_KEY=`
`$ aws s3api --endpoint http://localhost:9878 create-bucket --bucket bucket1`

_An error occurred (AccessDenied) when calling the CreateBucket operation: User doesn't have the right to access this resource._

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 1]
```diff
@@ -273,15 +278,16 @@ public UserGroupInformation createUGI() {
     if (userGroupInformation != null) {
       return userGroupInformation;
     }
-
     if (omRequest.hasUserInfo() &&
         !StringUtils.isBlank(omRequest.getUserInfo().getUserName())) {
       userGroupInformation = UserGroupInformation.createRemoteUser(
           omRequest.getUserInfo().getUserName());
+      LOG.debug("creating UGI remote user for acl");
```
[the remark this answers is not in the document]
Comment (neils-dev) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java]: Updated with latest commit to omit log statement.  Thanks.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 2]
```diff
@@ -273,15 +278,16 @@ public UserGroupInformation createUGI() {
...
       return userGroupInformation;
     } else {
       // This will never happen, as for every OM request preExecute, we
       // should add userInfo.
+      LOG.debug("NO UGI for acl");
```
[the remark this answers is not in the document]
Comment (neils-dev) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java]: Also updated with latest commit to omit log statement.  Thanks.

**Review by bharatviswa504 — COMMENTED** [review]
@neils-dev Thanks for addressing the issue.
Can you share info on my comment on UGI thing, how it is being populated.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 3]
```diff
@@ -4038,7 +4038,7 @@ public ResolvedBucket resolveBucketLink(Pair<String, String> requested)
       if (isAclEnabled) {
         InetAddress remoteIp = Server.getRemoteIp();
         resolved = resolveBucketLink(requested, new HashSet<>(),
-            Server.getRemoteUser(),
+            getRemoteUser(),
```
[the remark this answers is not in the document]
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4041]: @neils-dev ??

**Review by avijayanhwx — COMMENTED** [review]
Thanks for working on this @kerneltime. I am +1 on this change. 

Just something to keep in mind, we need to make sure the error codes for s3 request failures (wrong creds, OM down etc) are semantically correct and consistent in both the RPC and GRPC implementations. For example, we now (incorrectly) return a 500 error for wrong credentials.

**Comment by bharatviswa504** [comment]
Thank You @neils-dev for the offline discussion and explaination.

**Pull request merged**